### PR TITLE
Unit tests

### PR DESCRIPTION
**Pull request by Aqualex**
added unit tests for iex feed 
still need to add unit tests for twap and vwap

### CODE REVIEW

File: UnitTesting/1unittest.q [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+//load files
+\l k4unit.q
+
+benchCheck:{$[`quoteBENCHMARK.csv in key def[`benchM];
+              $[`tradeBENCHMARK.csv in key def[`benchM];
+                $[`tablerBENCHMARK.csv in key def[`benchM];
+                 [0];
+                 -1"tablerBENCHMARK.csv MISSING"];
+              -1"tradeBENCHMARK.csv MISSING"];
+            -1"BENCHMARKS MISSING"]};
```
Comment (ThomasSmyth) [inline UnitTesting/1unittest.q]: Fix this.

Comment (Aqualex) [reply UnitTesting/1unittest.q]: please look at the new file in TorQ-Finance > UnitTesting > unitTests.q

Comment (Aqualex) [reply UnitTesting/1unittest.q]: also the unit tests (the csv file) can be found in TorQ-Finance-Starter-Pack > UnitTesting > utests > 1iexfeed.csv 
apologies for not being clear about this

Comment (jonathonmcmurray) [reply UnitTesting/1unittest.q]: If this file isn't meant to be here, make sure it isn't here

Comment (Aqualex) [reply UnitTesting/1unittest.q]: I will remove the files first thing in the morning!

File: UnitTesting/k4unit.q [hunk 2]
```diff
@@ -0,0 +1,143 @@
+/ k4 unit testing, loads tests from csv's, runs+logs to database
```
Comment (jonathonmcmurray) [inline UnitTesting/k4unit.q]: this script is already in TorQ repo, no need to add it here as well, you just might need to update a few paths to use the existing copy

File: UnitTesting/runtests.q [hunk 3]
```diff
@@ -0,0 +1,8 @@
+KUltd each hsym`$.proc.params[`test]
```
Comment (jonathonmcmurray) [inline UnitTesting/runtests.q]: As with `k4unit.q`, already in TorQ repo

File: appconfig/settings/iexfeed.q [hunk 4]
```diff
@@ -9,13 +9,19 @@ CONNECTIONS:enlist `tickerplant         // Feedhandler connects to the tickerpla
 HOPENTIMEOUT:30000
```
Comment (jonathonmcmurray) [inline appconfig/settings/iexfeed.q]: this file is unrelated to unit tests, shouldn't be in this PR

File: appconfig/settings/quote_iex.csv [hunk 5]
```diff
@@ -0,0 +1,9 @@
+ocol,ncol,typ
```
Comment (jonathonmcmurray) [inline appconfig/settings/quote_iex.csv]: this file is unrelated to unit tests, shouldn't be in this PR

File: appconfig/settings/trade_iex.csv [hunk 6]
```diff
@@ -0,0 +1,8 @@
+ocol,ncol,typ
```
Comment (jonathonmcmurray) [inline appconfig/settings/trade_iex.csv]: this file is unrelated to unit tests, shouldn't be in this PR

File: code/iexfeed/iex.q [hunk 7]
```diff
@@ -1,62 +1,92 @@
 \d .iex
```
Comment (jonathonmcmurray) [inline code/iexfeed/iex.q]: this file is unrelated to unit tests, shouldn't be in this PR

File: UnitTesting/unitTests.q [hunk 8]
```diff
@@ -0,0 +1,50 @@
+//DEFAULT VALUES
+def:.Q.def[`stackID`user`pass`testCSV!(9000;`admin;`admin;`:utests/1iexfeed.csv)].Q.opt[.z.x]
```
Comment (jonathonmcmurray) [inline UnitTesting/unitTests.q]: logging in as admin is a no-no. also, if you load this as a TorQ process no need to do this, it'll do it for you in `.proc.params`

**Comment by ThomasSmyth** [comment]
`k4unit.q` is in TorQ, does that need removed too?

File: UnitTesting/unitTests.q [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+
+//LOADING Q-SCRIPTS
+\l k4unit.q
+
+//UTILITIES
+//get the right port to open handle 
+getP:{[proc] 
+     $[`rdb~proc;string[def[`stackID]+2];::]
+     $[`feed~proc;string[def[`stackID]+14];::]
+     $[`iex~proc;string[def[`stackID]+19];::]};
```
Comment (ThomasSmyth) [inline UnitTesting/unitTests.q]: Can you pull these from the config file instead? + lack of semi colons here could cause unexpected behaviour.

File: UnitTesting/unitTests.q [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+
+//UTILITIES
+//get the right port to open handle 
+getP:{[proc] 
+     $[`rdb~proc;string[def[`stackID]+2];::]
+     $[`feed~proc;string[def[`stackID]+14];::]
+     $[`iex~proc;string[def[`stackID]+19];::]};  
+
+//creating path for opening handle 
+path:{`$"::",getP[x],":",string[def[`user]],":",string[def[`pass]]};
```
Comment (ThomasSmyth) [inline UnitTesting/unitTests.q]: Trim some squares here + is there a better way to do this? Seems untidy. 

I think tests should have a different username + password so you can track their requests in usage logs (you have `admin:admin` as defaults).

File: UnitTesting/utests/1iexfeed.csv [hunk 11]
```diff
@@ -0,0 +1,19 @@
+action,ms,bytes,lang,code,repeat,minver,comment
+run,,,q,dH[`iex]("a:.iex.getdata[.iex.mainurl;.iex.quotesuffix[\"CAT\"]]"),,,downloads data
+run,,,q,dH[`iex]("b:.iex.getdata[.iex.mainurl;.iex.tradesuffix[\"CAT\"]]"),,,download data
+run,,,q,dH[`iex]("c:.iex.getdata[.iex.mainurl;.iex.quotesuffix[\"DOG\"]]"),,,downloads data 
+run,,,q,dH[`iex]("d:.iex.getdata[.iex.mainurl;.iex.tradesuffix[\"DOG\"]]"),,,downloads data
+true,,,q,10h~type dH[`iex]("a"),,,checks if stremead data is of type 10h
+true,,,q,10h~type dH[`iex]("b"),,,checks if streamed data is of type 10h
+run,,,q,98h~type dH[`iex](".iex.createtable[`.iex.dqte;enlist .j.k a]"),,,check if function create table returns the right type 
+run,,,q,98h~type dH[`iex](".iex.createtable[`.iex.dtrd;.j.k b]"),,,check if functien create table returns the right type
```
Comment (jonathonmcmurray) [inline UnitTesting/utests/1iexfeed.csv]: checking the type is insufficient IMO, doesn't make sure that it's creating the *right* table

File: UnitTesting/unitTests.q [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+
+//function that checks operating system
+//can be ammended to use different system commands for different opSystems
+opSystem:{[command]$[`l64~.z.o; $[`list~command;show system"ls";::];"nyi"]};
+
+//timer.remove function does nothing if the feed is already removed from the table
+stFeed:{x(".timer.remove[first exec id from .timer.timer where {[x](`",string[y],"`)~x}'[funcparam]]")}; 
+
+//check if csv file or directory 
+loadTest:{$["csv"~-3#string[def[`testCSV]];@[KUltf;hsym def[`testCSV];{-2"ERROR: "x}];@[KUltd;hsym def[`testCSV];{-2"ERROR: ",x}]]};
```
Comment (ThomasSmyth) [inline UnitTesting/unitTests.q]: Why not just ``def[`testCSV]like"*.csv"``? To avoid using `hsym` multiple times here you could for `testCSV` to be a proper filepath when you handle the inputs.

File: UnitTesting/unitTests.q [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+opSystem:{[command]$[`l64~.z.o; $[`list~command;show system"ls";::];"nyi"]};
+
+//timer.remove function does nothing if the feed is already removed from the table
+stFeed:{x(".timer.remove[first exec id from .timer.timer where {[x](`",string[y],"`)~x}'[funcparam]]")}; 
+
+//check if csv file or directory 
+loadTest:{$["csv"~-3#string[def[`testCSV]];@[KUltf;hsym def[`testCSV];{-2"ERROR: "x}];@[KUltd;hsym def[`testCSV];{-2"ERROR: ",x}]]}; 
+
+//openning handle to process
+opHandle:{[pTO]@[hopen;path[pTO];{-2"ERROR: ",x}]}; //open handle to IEX feed
```
Comment (ThomasSmyth) [inline UnitTesting/unitTests.q]: No need for squares with path.

File: UnitTesting/unitTests.q [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+       dH::()!();
+       -1"OPENING HANDLES..."; 
+       dH[`rdb]::opHandle[`rdb];
+       dH[`iex]::opHandle[`iex]; 
+       dH[`feed]::opHandle[`feed];
+       -1"STOPING FEEDS..."; 
+       stFeed[dH[`iex];`.iex.timer]; 
+       stFeed[dH[`feed];`feed];
+       -1"RUNNING TESTS...";
+       KUrt[];
```
Comment (ThomasSmyth) [inline UnitTesting/unitTests.q]: Lots of extra square brackets here.

Comment (Aqualex) [reply UnitTesting/unitTests.q]: I prefer having the square brackets. I find it a lot easier to read the code

File: UnitTesting/unitTests.q (hunk 13 above)
Comment (Aqualex) [reply UnitTesting/unitTests.q]: I am not sure I understand what you mean ... opHandle:{[pTO]@[hopen;path`pTO;{-2"ERROR: ",x}]}; .. like this?

Comment (jonathonmcmurray) [reply UnitTesting/unitTests.q]: no backtick

```
opHandle:{[pTO]@[hopen;path pTO;{-2"ERROR: ",x}]};
```

File: UnitTesting/unitTests.q (hunk 12 above)
Comment (Aqualex) [reply UnitTesting/unitTests.q]: I did not think of it. I'll change this right now

File: UnitTesting/unitTests.q (hunk 10 above)
Comment (Aqualex) [reply UnitTesting/unitTests.q]: squares are a personal preference
I'll think about a better way of doing it 
User and Password can be specified as command line arguments (at the moment these are the details for connecting to our TorQ processes (at least in my testing env))

File: UnitTesting/unitTests.q (hunk 9 above)
Comment (Aqualex) [reply UnitTesting/unitTests.q]: I though about that as well. I will focus now on getting the tests right. This will be part of a future release. Thank you for pointing that out!

File: UnitTesting/utests/1iexfeed.csv (hunk 11 above)
Comment (Aqualex) [reply UnitTesting/utests/1iexfeed.csv]: I agree with that. I will create a dummy dictionary, run the function and compare the results.

File: UnitTesting/unitTests.q (hunk 8 above)
Comment (Aqualex) [reply UnitTesting/unitTests.q]: I thought of having unit tests as a standalone process that will be run on a copy of TorQ before bringing it back on-line.

File: code/iexfeed/iex.q (hunk 7 above)
Comment (Aqualex) [reply code/iexfeed/iex.q]: that is not my file. I haven't touched that.

File: appconfig/settings/trade_iex.csv (hunk 6 above)
Comment (Aqualex) [reply appconfig/settings/trade_iex.csv]: I am not sure I understand your comment. that has not been added by me ... or was it?

File: UnitTesting/runtests.q (hunk 3 above)
Comment (Aqualex) [reply UnitTesting/runtests.q]: I will update the paths and remove them from this directory. Thank you for pointing this out.

File: UnitTesting/k4unit.q (hunk 2 above)
Comment (Aqualex) [reply UnitTesting/k4unit.q]: I will update the paths and remove the file

File: appconfig/settings/trade_iex.csv (hunk 6 above)
Comment (jonathonmcmurray) [reply appconfig/settings/trade_iex.csv]: Yes, it is in your branch (which it shouldn't be) and therefore in your PR (which it shouldn't be)

Most likely what happened is you branched from the branch where iexfeed modifications were happening, instead of branching from master branch

You can either revert the various commits related to iexfeed, or maybe do some rebasing. But basically, all the iex changes should not be in this PR, and will cause problems when it comes to merging this & other PR both to master

File: code/processes/vtwap.q [hunk 15]
```diff
@@ -0,0 +1,65 @@
+tab:flip`time`sym`price`size`stop`cond`ex!();
```
Comment (jonathonmcmurray) [inline code/processes/vtwap.q]: this file shouldn't be in Unit Tests PR I don't think

File: code/iexfeed/iex.q [hunk 16]
```diff
@@ -14,7 +14,7 @@ timerperiod:@[value;`timerperiod;0D00:00:02.000];
 lvcq:@[value;`lvcq;1!flip `sym`bid`ask`bsize`asize`mode`ex`srctime!()];
 lvct:@[value;`lvct;1!flip`sym`price`size`stop`cond`ex`srctime!()];
 qcols:@[value;`qcols;`bid`ask`bsize`asize`mode`ex];
-nullq:@[value;`nullq;qcols!(0f;0f;0;0;" ";" ")];
+nullq:@[value;`nullq;qcols!(0f;0f;0i;0i;" ";" ")];
```
Comment (jonathonmcmurray) [inline code/iexfeed/iex.q]: this change isn't unit tests related, shouldn't be in this PR

File: appconfig/process.csv [hunk 17]
```diff
@@ -20,3 +20,4 @@ localhost,{KDBBASEPORT}+17,sortslave,sortslave2
 localhost,{KDBBASEPORT}+18,metrics,metrics1
 localhost,{KDBBASEPORT}+19,iexfeed,iexfeed1
 localhost,{KDBBASEPORT}+20,checker,symcheck
+localhost,{KDBBASEPORT}+21,vtwap,vtwap1
```
Comment (jonathonmcmurray) [inline appconfig/process.csv]: Shouldn't be in this PR

File: appconfig/passwords/vtwap.txt [hunk 18]
```diff
@@ -0,0 +1 @@
+vtwap:pass
```
Comment (jonathonmcmurray) [inline appconfig/passwords/vtwap.txt]: Shouldn't be in this PR

File: appconfig/passwords/accesslist.txt [hunk 19]
```diff
@@ -17,3 +17,4 @@ metrics:pass
 vwapsub:pass
 checker:pass
 admin:admin
+vtwap:pass
```
Comment (jonathonmcmurray) [inline appconfig/passwords/accesslist.txt]: Shouldn't be in this PR

File: UnitTesting/utests/1iexfeed.csv [hunk 20]
```diff
@@ -1,19 +1,24 @@
...
-true,,,q,0=dH[`iex]("count .iex.checkdup[;;`.iex.lvcq;.iex.qcols;.iex.nullq]/[0#testTabDup;testTabDup]"),,,checking if checkdup works as it should
+run,,,q,dH[`iex]("tN::([]sym:`CAT`DOG;bid:(0f;0f); ask:(0f;0f); bsize:(0i;0i); asize:(0i;0i);mode:(\" \";\" \"); ex:(\" \";\" \");srctime:2001.01.01D01:01:01.100000000)"),,,create test table for nulls
+true,,,q,0=dH[`iex]("count .iex.checkdup[;;`.iex.lvcq;.iex.qcols;.iex.nullq]/[0#tN;tN]"),,,check if count 0
+run,,,q,dH[`iex]("tD::([]sym:enlist `CAT;bid:120f;ask:140f;bsize:100i;asize:200i;mode:\"T\";ex:\"T\";srctime:2001.01.01D01:01:01.100000000)"),,,prepare test table
+run,,,q,dH[`iex]("delete from `.iex.lvcq;"),,,clear last memory cached 
+run,,,q,dH[`iex]("`.iex.lvcq upsert tD"),,,upsert testTabDup in lvcq to record it as a duplicate 
+true,,,q,0=dH[`iex]("count .iex.checkdup[;;`.iex.lvcq;.iex.qcols;.iex.nullq]/[0#tD;tD]"),,,checking if checkdup works as it should
+true,,,q,vwapBENCH~{(dH[`vtwap]("getvwap"))[x;16:30;17:00]}'[key .vtwap.data],,,compare vwap with Benchmark
```
Comment (jonathonmcmurray) [inline UnitTesting/utests/1iexfeed.csv]: file name suggests this is unit tests for iex, not [vt]wap subscriber
